### PR TITLE
Revised .NET Docker Images Topic and Created two new Docker Topics

### LINKED ISSUE #2652: Revise Building Docker Images for .NET Core Applications Topic
# Revise Building Docker Images for .NET Core Applications Topic

This page needs to be revised: [Building .NET docker Images](https://docs.microsoft.com/en-us/dotnet/articles/core/docker/building-net-docker-images)

This will take the place of Issue #2035 - Update .NET Core Docker Doc 

## Needed Revisions
- [x] Move to the .NET Guide in the Table of Contents
- [x] Change Title to "Introduction to .NET and Docker"
- [x] Revised Outline
   - [x] Introduction
      - [x] What is Docker? 
      - [x] Further Reading.
    - [x]  .NET Docker Images
       - [x] Scenario Based Guidance
           - [x] Explain different OS base images better. Link to more Docker documentation on the different Docker clients (Mac, Linux, Windows).
       - [x] Cover scenarios: .net core, .net fx, aspnet core, aspnet fx, wcf, iis
       - [x] Interact with other msft stack images: mssql, vsts-agent, oms, azure-cli, etc. 
         - [x] Add multi-stage build explanation for using both sdk and runtime images in Dockerfile
    - [x] Explore the Rich Docker ecosystem. Link to microservices doc.
    - [x] Further Reading on .NET Core images: production, development/build explanations. 
        - [x] Add links to aspnetcore images.
        - [x] Provide links at bottom for related .NET Framework images doc
- [x] Develop with Docker
       - [ ] Describe what gaps our tools bring to developing with Docker
       - [ ]  Debug through a container in VS! Link to P2 tutorial.
      -  [ ] Getting Started with Docker and VS4Mac

Note: Consider splitting Image documentation into a separate doc than VS Code and Docker use.

### LINKED ISSUE #2683: [NewTopic] Learn Docker Basics with .NET Core
# [NewTopic] Learn Docker Basics with .NET Core
1. Table of Contents.
  .NET Core Guide -> Docker
2. Abstract
  Learn about Docker basic commands using .NET Core. This doc should be OS agnostic.
3. Outline
- [x] Introduction   
  - [x] .NET Core: easiest way to get started.
  - [x]  Briefly mention windows and linux. Link to [announcement](https://github.com/dotnet/announcements/issues/14) for more multi-arch info.
- [x] Installing Docker Client
      - [ ] Download Docker client (Either Windows, Linux, or Mac)
- [x] Essential Docker commands  
      - [ ] build
      - [ ] run
      - [ ] ps
      - [ ] stop
      - [ ] rm
      - [ ] rmi
- [x] Your first .NET Core Docker app
    - [x] Hello console app - Walkthrough of creating the .NET Core Hello World application to have something to Dockerize.
    - [x] Dockerize the Hello console app - Walkthrough of dockerizing the .NET Core Hello World application .
      - [x]  Inside a Dockerfile - Walkthrough of the simple Dockerfile.
      - [x] Create the Hello app Dockerfile 
      - [x] Build and run the Hello Docker app. 
- [x] Next Steps
    - [x] Debugging a container with Visual Studio Docker Tools
    - [x] Link to [aspnetcore DockerHub repo] or tutorial.

### PR DESCRIPTION
**Pull request by JRAlexander**
# Revised .NET Docker Images Topic and Created two new Docker Topics

## Summary
Updated "Building Docker Images for .NET Core Applications" Topic, and created two new topics - "Introduction to .NET and Docker", and "Learn Docker Basics with .NET Core ".

[Building Docker Images for .NET Core Applications - Internal Review URL](https://review.docs.microsoft.com/en-us/dotnet/core/docker/building-net-docker-images?branch=pr-en-us-3380)
[Introduction to .NET and Docker - Internal Review URL](https://review.docs.microsoft.com/en-us/dotnet/core/docker/intro-net-docker?branch=pr-en-us-3380)
[Learn Docker Basics with .NET Core - Internal Review URL](https://review.docs.microsoft.com/en-us/dotnet/core/docker/docker-basics-dotnet-core?branch=pr-en-us-3380)

Fixes #2683 
Fixes #2652 

This PR completes #2652 Revise Building Docker Images for .NET Core Applications Topic, and
#2683 [NewTopic] Learn Docker Basics with .NET Core.

## Details
Please see the above issues for more details. 

## Suggested Reviewers
@kendrahavens, @mairaw

### CODE REVIEW

File: docs/core/docker/building-net-docker-images.md [hunk 1]
```diff
@@ -2,224 +2,251 @@
 title: Building .NET Core Docker Images
 description: Understanding Docker images and .NET Core
 keywords: .NET, .NET Core, Docker
-author: spboyer
-ms.author: shboyer
-ms.date: 09/06/2017
+author: johalex
```
Comment (mairaw) [inline docs/core/docker/building-net-docker-images.md]: I'll try to review your PR shortly. your GH alias is JRAlexander

File: docs/core/docker/building-net-docker-images.md [hunk 2]
```diff
@@ -2,224 +2,251 @@
...
+   Connection-specific DNS Suffix  . : contoso.com
+   Link-local IPv6 Address . . . . . : fe80::1967:6598:124:cfa3%4
+   IPv4 Address. . . . . . . . . . . : 172.29.245.43
+   Subnet Mask . . . . . . . . . . . : 255.255.240.0
+   Default Gateway . . . . . . . . . : 172.29.240.1
 ```
 
-The command creates the image based on the smaller **microsoft/dotnet:core** base image, [EXPOSE](https://docs.docker.com/engine/reference/builder/#/expose) port 5000, sets the [ENTRYPOINT](https://docs.docker.com/engine/reference/builder/#/entrypoint) for `dotnet api.dll` and copies it to the `/app` directory. There is no debugger, SDK or `dotnet restore` ([see note](#dotnet-restore-note)) resulting in a much smaller image. The image is named **api** with a `TAG` of **latest**.
+> [!Note]
+> Docker exec runs a new command in a running command. For more information, see the [docker exec reference](https://docs.docker.com/engine/reference/commandline/exec/) on command-line parameters.
```
Comment (BillWagner) [inline docs/core/docker/building-net-docker-images.md]: typo: "in a running command" in the first sentence should be "in a running container."

File: docs/core/docker/building-net-docker-images.md [hunk 3]
```diff
@@ -2,224 +2,251 @@
...
 ```
 
-Another way to generate the image and run the application within the Docker container is to open the application in Visual Studio Code and use the debugging tools. 
+> [!IMPORTANT]
+> You must navigate to the **container IP address** (as opposed to http://localhost) in your browser directly when using Windows containers. You can get the IP address of your container with the following steps:
 
-Select the debugging icon in the View Bar on the left side of Visual Studio Code.
+* Open up another command prompt.
+* Run `docker ps` to see your running containers. The "aspnetcore_sample" container should be there.
+* Run `docker exec` aspnetcore_sample ipconfig.
```
Comment (BillWagner) [inline docs/core/docker/building-net-docker-images.md]: the 2nd back-tick should be after 'ipconfig', not after 'exec'

File: docs/core/docker/building-net-docker-images.md [hunk 4]
```diff
@@ -2,224 +2,251 @@
...
+You can produce an application that is ready to deploy to production locally using the [dotnet publish](../tools/dotnet-publish.md) command.
+
+```console
+dotnet publish -c release -o published
+```
+
+> [!Note]
+> The -c release argument builds the application in release mode (the default is debug mode). For more information, see the [dotnet run reference](../tools/dotnet-run.md) on command-line parameters.
+
+You can run the application on **Windows** using the `dotnet published` command.
```
Comment (BillWagner) [inline docs/core/docker/building-net-docker-images.md]: "published" isn't a .NET CLI verb. Rather, you are telling the .NET CLI to run your application DLL from the publish directory.

File: docs/core/docker/building-net-docker-images.md [hunk 5]
```diff
@@ -2,224 +2,251 @@
...
+
+```console
+dotnet published\aspnetapp.dll
 ```
-REPOSITORY          TAG                  IMAGE ID            CREATED             SIZE
-api                 debug                70e89fbc5dbe        1 hour ago        779.8 MB
-api                 latest               ef17184c8de6        1 hour ago        260.7 MB
+
+You can run the application on **Linux** or **macOS** using the `dotnet published` command.
```
Comment (BillWagner) [inline docs/core/docker/building-net-docker-images.md]: previous comment applies here as well.

File: docs/core/docker/docker-basics-dotnet-core.md [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+
+## Your first .NET Core Docker app
+
+### Prerequisites
+
+#### .NET Core 2.0 SDK
+
+* Install [.NET Core SDK 2.0](https://www.microsoft.com/net/core).
+
+* Install your favorite code editor, if you haven't already.
```
Comment (BillWagner) [inline docs/core/docker/docker-basics-dotnet-core.md]: I'd recommend Visual Studio and VS Code here, if a reader doesn't have a favorite editor.

File: docs/core/docker/docker-basics-dotnet-core.md [hunk 7]
```diff
@@ -0,0 +1,280 @@
...
+
+   * [Fedora](https://www.docker.com/docker-fedora)
+
+   * [Ubuntu](https://www.docker.com/docker-ubuntu)
+
+* [macOS](https://docs.docker.com/docker-for-mac/)
+
+* [Windows](https://docs.docker.com/docker-for-windows/).
+
+### Create a .NET Core 2.0 console app for Dockerization
```
Comment (BillWagner) [inline docs/core/docker/docker-basics-dotnet-core.md]: We have so many of these "get started" bits, can we link to one? /cc @mairaw

File: docs/core/docker/docker-basics-dotnet-core.md [hunk 8]
```diff
@@ -0,0 +1,280 @@
...
+
+Open your text editor and let's get started! We're still working from the Hello directory we built the app in.
+
+Add the following Docker instructions for either Linux or [Windows Containers](https://docs.microsoft.com/en-us/virtualization/windowscontainers/about/) to a new file. When finished, save it in the root of your Hello directory as **Dockerfile**, with no extension (You may need to set your file type to `All types (*.*)` or something similar).
+
+```Dockerfile
+FROM microsoft/dotnet:2.0-sdk
+WORKDIR /app
+
+# copy csproj and restore as distinct layers
```
Comment (BillWagner) [inline docs/core/docker/docker-basics-dotnet-core.md]: Is this still needed, now that `dotnet publish` executes `dotnet restore`?

File: docs/core/docker/docker-basics-dotnet-core.md [hunk 9]
```diff
@@ -0,0 +1,280 @@
...
+* [docker build](https://docs.docker.com/engine/reference/commandline/build/)
+* [docker run](https://docs.docker.com/engine/reference/commandline/run/)
+* [docker ps](https://docs.docker.com/engine/reference/commandline/ps/)
+* [docker stop](https://docs.docker.com/engine/reference/commandline/stop/)
+* [docker rm](https://docs.docker.com/engine/reference/commandline/rm/)
+* [docker rmi](https://docs.docker.com/engine/reference/commandline/rmi/)
+
+#### Build and run
+
+We wrote the dockerfile; now Docker builds our app and then runs the container.
```
Comment (BillWagner) [inline docs/core/docker/docker-basics-dotnet-core.md]: Try to use 2nd person: You instead of We

File: docs/core/docker/docker-basics-dotnet-core.md [hunk 10]
```diff
@@ -0,0 +1,280 @@
...
+
+Here are some next steps you can take:
+
+* [Introduction to .NET Docker Images Video](https://channel9.msdn.com/Shows/Code-Conversations/Introduction-to-NET-Docker-Images-with-Kendra-Havens?term=docker)
+* [Visual Studio, Docker & Azure Container Instances better together!](https://blogs.msdn.microsoft.com/alimaz/2017/08/17/visual-studio-docker-azure-container-instances-better-together/)
+* [Docker for Azure Quickstarts](https://docs.docker.com/docker-for-azure/#docker-community-edition-ce-for-azure)
+* [Deploy your app on Docker for Azure](https://docs.docker.com/docker-for-azure/deploy/)
+
+> [!Note]
+> If you do not have an Azure subscription, [sign up today](https://azure.microsoft.com/en-us/free/?b=16.48) for a free 30-day account and get $200 in Azure Credits to try out any combination of Azure services.
```
Comment (BillWagner) [inline docs/core/docker/docker-basics-dotnet-core.md]: remove en-us for localization

File: docs/core/docker/docker-basics-dotnet-core.md [hunk 11]
```diff
@@ -0,0 +1,280 @@
...
+## Docker Images used in this sample
+
+The following Docker images are used in this sample
+
+* [`microsoft/dotnet:2.0-sdk`](https://hub.docker.com/r/microsoft/dotnet)
+
+## Related Resources
+
+* [.NET Core Docker samples](https://github.com/dotnet/dotnet-docker-samples/README.md)
+* [Dockerfile on Windows Containers](https://docs.microsoft.com/en-us/virtualization/windowscontainers/manage-docker/manage-windows-dockerfile)
```
Comment (BillWagner) [inline docs/core/docker/docker-basics-dotnet-core.md]: remove en-us

File: docs/core/docker/docker-basics-dotnet-core.md [hunk 12]
```diff
@@ -0,0 +1,280 @@
...
+* [.NET Framework Docker samples](https://github.com/Microsoft/dotnet-framework-docker-samples)
+* [ASP.NET Core on DockerHub](https://hub.docker.com/r/microsoft/aspnetcore/)
+* [Dockerize a .NET Core application - Docker Tutorial](https://docs.docker.com/engine/examples/dotnetcore/)
+* [Working with Visual Studio Docker Tools](https://docs.microsoft.com/en-us/aspnet/core/publishing/visual-studio-tools-for-docker)
```
Comment (BillWagner) [inline docs/core/docker/docker-basics-dotnet-core.md]: remove en-us

File: docs/core/docker/intro-net-docker.md [hunk 13]
```diff
@@ -0,0 +1,200 @@
...
+ 
+The [Docker platform](https://docs.docker.com/engine/docker-overview/#the-docker-platform) uses the [Docker Engine](https://docs.docker.com/engine/docker-overview/#docker-engine) to quickly build and package apps as [Docker images](https://docs.docker.com/glossary/?term=image) written in the [Dockerfile](https://docs.docker.com/glossary/?term=Dockerfile) [format that then is deployed and run in a [layered container](https://docs.docker.com/engine/userguide/storagedriver/imagesandcontainers/#container-and-layers).
+
+You can either create your own [layered images](https://docs.docker.com/engine/userguide/storagedriver/imagesandcontainers/#images-and-layers) as dockerfiles or use existing ones from a [registry](https://docs.docker.com/glossary/?term=registry), like [Docker Hub](https://docs.docker.com/glossary/?term=Docker%20Hub).
```
Comment (BillWagner) [inline docs/core/docker/intro-net-docker.md]: I'd rephrase this: you don't create an image **as** a dockerfile, but the image is created **using** a dockerfile.

File: docs/core/docker/intro-net-docker.md [hunk 14]
```diff
@@ -0,0 +1,200 @@
...
+
+The [relationship between Docker containers, images, and registries](https://docs.microsoft.com/en-us/dotnet/standard/microservices-architecture/container-docker-introduction/docker-containers-images-registries) is an important concept when [architecting and building containerized applications or microservices](https://docs.microsoft.com/en-us/dotnet/standard/microservices-architecture/architect-microservice-container-applications/). This approach greatly shortens the time between development and deployment.
```
Comment (BillWagner) [inline docs/core/docker/intro-net-docker.md]: remove en-us from both links

File: docs/core/docker/intro-net-docker.md [hunk 15]
```diff
@@ -0,0 +1,200 @@
...
+
+### Further reading (and watching)
+
+* [Windows-based containers: Modern app development with enterprise-grade control.](https://www.youtube.com/watch?v=Ryx3o0rD5lY&feature=youtu.be)
+* [Docker overview](https://docs.docker.com/engine/docker-overview/)
+* [Dockerfile on Windows Containers](https://docs.microsoft.com/en-us/virtualization/windowscontainers/manage-docker/manage-windows-dockerfile)
```
Comment (BillWagner) [inline docs/core/docker/intro-net-docker.md]: remove en-us

File: docs/core/docker/intro-net-docker.md [hunk 16]
```diff
@@ -0,0 +1,200 @@
...
+* [Best practices for writing Dockerfiles](https://docs.docker.com/engine/userguide/eng-image/dockerfile_best-practices/)
+* [Building Docker Images for .NET Core applications](..\building-net-docker-images.md)
```
Comment (BillWagner) [inline docs/core/docker/intro-net-docker.md]: use '/' instead of '\'

File: docs/core/docker/intro-net-docker.md [hunk 17]
```diff
@@ -0,0 +1,200 @@
...
+
+* **Develop in a container** — You can develop in a consistent environment, making development and production environments similar (avoiding issues like global state on developer machines). Visual Studio Tools for Docker even enable you to start a container directly from Visual Studio.
+
+* **Test in a container** — You can test in a container, reducing failures due to incorrectly configured environments or other changes left behind from the last test.
+
+* **Build in a container** — You can build code in a container, avoiding the need to correctly configure shared build machines for multiple environments but instead move to a “BYOC” (bring your own container) approach.
+
+* **Deployment in all environments** — You can deploy an image through all of your environments. This approach reduces failures due to configuration differences, typically changing the image behavior via external configuration (for example, injected environment variables).
+
+[General guidance](https://docs.microsoft.com/en-us/dotnet/standard/microservices-architecture/net-core-net-framework-containers/general-guidance) for deciding between .NET Core and .NET Framework for Docker container development.
```
Comment (BillWagner) [inline docs/core/docker/intro-net-docker.md]: remove en-us

File: docs/core/docker/intro-net-docker.md [hunk 18]
```diff
@@ -0,0 +1,200 @@
...
+
+* [.NET Core images on DockerHub](https://hub.docker.com/r/microsoft/dotnet/)
+
+* [Dockerize a .NET Core application](https://docs.docker.com/engine/examples/dotnetcore/)
+
+* This .NET Core Docker sample demonstrates how to [use Docker in your .NET Core development process](https://github.com/dotnet/dotnet-docker-samples/tree/master/dotnetapp-dev). The sample works with both Linux and Windows containers.
+
+* This .NET Core Docker sample demonstrates a best practice pattern for [building Docker images for .NET Core apps for production.](https://github.com/dotnet/dotnet-docker-samples/tree/master/dotnetapp-prod) The sample works with both Linux and Windows containers.
+
+* This [.NET Core Docker sample](https://github.com/dotnet/dotnet-docker-samples/tree/master/dotnetapp-selfcontained) demonstrates a best practice pattern for building Docker images for [self-contained .NET Core applications](https://docs.microsoft.com/en-us/dotnet/core/deploying/). Used for the smallest production container without a benefit from [sharing base images between containers](https://docs.docker.com/engine/userguide/storagedriver/imagesandcontainers/). However, lower Docker layers could be shared.
```
Comment (BillWagner) [inline docs/core/docker/intro-net-docker.md]: remove en-us from links.

There are several more below in this same file. All should be removed.

**Review by BillWagner — CHANGES_REQUESTED** [review]
Overall, this is great content.  I found a number of small items to update.

File: docs/core/docker/building-net-docker-images.md (hunk 1 above)
Comment (JRAlexander) [reply docs/core/docker/building-net-docker-images.md]: Thanks! Good catch

File: docs/core/docker/intro-net-docker.md (hunk 13 above)
Comment (JRAlexander) [reply docs/core/docker/intro-net-docker.md]: Great point.

File: docs/core/docker/building-net-docker-images.md [hunk 19]
```diff
@@ -2,224 +2,251 @@
...
+    For example, the `dotnet publish` output contains:
+
+    * the compiled binaries
+    * .js and .css files
 
-Though there are multiple versions of the .NET Core image, they all share one or more layers. The amount of disk space needed to store or the delta to pull from your registry is much smaller than the whole because all of the images share the same base layer and potentially others.  
+
+The reason to include the `dotnet publish` command output in your production image is to keep its' size to a minimum.
+
+Though there are multiple versions of the .NET Core image, they all share one or more layers. This architecture decreases the needed disk space.
```
[the remark this answers is not in the document]
Comment (JRAlexander) [reply docs/core/docker/building-net-docker-images.md]: I believe that's correct.

File: docs/core/docker/docker-basics-dotnet-core.md (hunk 7 above)
Comment (JRAlexander) [reply docs/core/docker/docker-basics-dotnet-core.md]: I think it would break the flow.

File: docs/core/docker/docker-basics-dotnet-core.md (hunk 8 above)
Comment (JRAlexander) [reply docs/core/docker/docker-basics-dotnet-core.md]: It's from the .NET Docker Sample Dockerfile.

**Comment by JRAlexander** [comment]
Having to close and reopen due to build errors

File: docs/core/docker/intro-net-docker.md (hunk 18 above)
Comment (JRAlexander) [reply docs/core/docker/intro-net-docker.md]: @BillWagner - All changes have been addressed and content rebuilt. Thanks for a thorough and thoughtful review!!

**Comment by JRAlexander** [comment]
@BillWagner - All changes have been addressed.

**Review by BillWagner — APPROVED** [review]

**Comment by BillWagner** [comment]
@JRAlexander The updates look good to me. :shipit: when you are happy with the product team feedback.

**Review by spboyer — APPROVED** [review]
`docker exec` is not the best practice for getting networking information about the container.  However in this case, just getting the ip it works just fine. A universal command is `docker inspect --format '{{ .NetworkSettings.Networks.nat.IPAddress }}' <container>` 

The article [Visual Studio Tools for Docker](https://review.docs.microsoft.com/en-us/aspnet/core/publishing/visual-studio-tools-for-docker) is missing or a dead link and needs updating for the multi-stage builds explanation.  The current doc uses VS 2015 and we need to push towards 2017.

All of the other docs look good.  The list of Docker images are `dotnet` but the examples are for ASP.NET. Might be important to show the microsoft/aspnet images as well and add the variations such as `-stretch`, `-sdk`, `-runtime` etc and why they are there OR discuss these in an ASP.NET specific document on the ASP.NET docs/deployments?

**Comment by JRAlexander** [comment]
Thanks, @spboyer for the feedback and review! I will pass that info onto the ASP.NET team, unless you'd like to create the article and add a PR.

**Comment by kendrahavens** [comment]
@MichaelSimons May also be interested in reviewing.

File: docs/core/docker/building-net-docker-images.md [hunk 20]
```diff
@@ -2,224 +2,255 @@
...
+docker build -t aspnetapp .
+docker run -it --rm --name aspnetcore_sample aspnetapp
 ```
 
-Another way to generate the image and run the application within the Docker container is to open the application in Visual Studio Code and use the debugging tools. 
+> [!IMPORTANT]
+> You must navigate to the **container IP address** (as opposed to http://localhost) in your browser directly when using Windows containers. You can get the IP address of your container with the following steps:
 
-Select the debugging icon in the View Bar on the left side of Visual Studio Code.
+* Open up another command prompt.
```
Comment (MichaelSimons) [inline docs/core/docker/building-net-docker-images.md]: These steps seem overly complicated.  The simple approach would be
`docker inspect --format '{{ .NetworkSettings.Networks.nat.IPAddress }}' aspnetcore_sample`

File: docs/core/docker/building-net-docker-images.md [hunk 21]
```diff
@@ -2,224 +2,255 @@
...
+* `microsoft/dotnet:<version>-runtime-deps`: The **runtime-deps** image contains the operating system with all of the native dependencies needed by .NET Core. This image is for [self-contained applications](https://docs.microsoft.com/dotnet/core/deploying/index).
 
 Latest versions of each variant:
 
-- `microsoft/dotnet` or `microsoft/dotnet:latest` (includes SDK)
-- `microsoft/dotnet:onbuild`
-- `microsoft/dotnet:core`
-- `microsoft/dotnet:core-deps`
+* `microsoft/dotnet` or `microsoft/dotnet:latest` (includes SDK)
```
Comment (MichaelSimons) [inline docs/core/docker/building-net-docker-images.md]: The latest tag is an alias for the sdk image.  Saying it includes the SDK is wrong - it is the sdk image.

File: docs/core/docker/building-net-docker-images.md [hunk 22]
```diff
@@ -2,224 +2,255 @@
...
-microsoft/dotnet    onbuild                 19b6a6c4b1db        540.4 MB
-microsoft/dotnet    1.0.0-preview2-sdk      a92c3d9ad0e7        540.4 MB
-microsoft/dotnet    core                    5224a9f2a2aa        253.2 MB
-microsoft/dotnet    1.0.0-core-deps         c981a2eebe0e        166.2 MB
-microsoft/dotnet    core-deps               c981a2eebe0e        166.2 MB
-microsoft/dotnet    latest                  03c10abbd08a        540.4 MB
-microsoft/dotnet    1.0.0-core              b8da4a1fd280        253.2 MB
+```console
+REPOSITORY          TAG                 IMAGE ID            SIZE
+microsoft/dotnet    latest              cc72b1545c62        1.64GB
```
Comment (MichaelSimons) [inline docs/core/docker/building-net-docker-images.md]: What is the motivation in stating the sizes of the images?  These tags are manifest backed which means the concrete image pulled differs based on the host platform (Linux/Windows, AMD64/ARM32).  This section makes no mention of this therefore the stated sizes are wrong depending on the users perspective.

File: docs/core/docker/building-net-docker-images.md [hunk 23]
```diff
@@ -2,224 +2,255 @@
...
 
-Though there are multiple versions of the .NET Core image, they all share one or more layers. The amount of disk space needed to store or the delta to pull from your registry is much smaller than the whole because all of the images share the same base layer and potentially others.  
+
+The reason to include the `dotnet publish` command output in your production image is to keep its' size to a minimum.
+
+Some .NET Core images share layers between different tags so downloading the latest tag is a relatively lightweight process. If you already have an older version on your machine, this architecture decreases the needed disk space.
+
+When multiple applications use common images on the same machine, memory is shared between the common images. The images must be the same to be shared.
 
 ## Docker image variations
```
Comment (MichaelSimons) [inline docs/core/docker/building-net-docker-images.md]: I think it is important to mention what platforms the images support.  Right now these images work on both AMD64 Windows (RS1 and RS3) and AMD64 Linux.  Linux AMD32V7 support is in preview.

File: docs/core/docker/docker-basics-dotnet-core.md [hunk 24]
```diff
@@ -0,0 +1,287 @@
...
+
+   `Program.cs`:
+
+   The program starts by `using System`. This statement means, "Bring everything in the `System` namespace into scope for this file." The `System` namespace includes basic constructs such as `string`, or numeric types.
+
+   We then define a namespace called `Hello`. You can change namespace to anything you want. A class named `Program` is defined within that namespace, with a `Main` method that takes an array of strings as its argument. This array contains the list of arguments passed in when the compiled program is called. In our example, the program only writes "Hello World!" to the console.
+
+2. `$ dotnet restore`
+
+   In .NET Core 2.x, **dotnet new** runs the [`dotnet restore`](../tools/dotnet-restore.md) command. **Dotnet restore** restores the tree of dependencies with a [NuGet](https://www.nuget.org/)(.NET package manager) call.
```
Comment (MichaelSimons) [inline docs/core/docker/docker-basics-dotnet-core.md]: `dotnet restore` is triggered by more than just `new`.  It is now a command that is implicitly invoked with any command that requires it.

File: docs/core/docker/docker-basics-dotnet-core.md (hunk 8 above)
Comment (MichaelSimons) [reply docs/core/docker/docker-basics-dotnet-core.md]: I think the subsequent text should be explaining why `dotnet restore` is done explicitly.  It is done to help improve the iterative development experience.  If you make a code tweak and rebuild, the layers up to `COPY . .` don't have to be rebuilt and docker will utilize the cached layers.    See the [Build Cache](https://docs.docker.com/engine/userguide/eng-image/dockerfile_best-practices/#build-cache) section of the Dockerfile best practices for more details.

File: docs/core/docker/docker-basics-dotnet-core.md [hunk 25]
```diff
@@ -0,0 +1,287 @@
...
+
+The [**ENTRYPOINT**](https://docs.docker.com/engine/reference/builder/#entrypoint) instruction allows the container to run as an executable.
+
+```Dockerfile
+ENTRYPOINT ["dotnet", "out/Hello.dll"]
+```
+
+Now you have a Dockerfile that:
+
+* copies your app to the container
```
Comment (MichaelSimons) [inline docs/core/docker/docker-basics-dotnet-core.md]: `image` not `container`.  An `image` is what is produced by building a Dockerfile.  A `container` is a running instance of an `image`.

File: docs/core/docker/docker-basics-dotnet-core.md [hunk 26]
```diff
@@ -0,0 +1,287 @@
...
+* your app's dependencies to the container
+* builds the app to run as an executable
+
+### Build and run the Hello .NET Core 2.0 app
+
+#### Essential Docker commands
+
+These Docker commands are essential:
```
Comment (MichaelSimons) [inline docs/core/docker/docker-basics-dotnet-core.md]: How was this list determined?  I am surprised a command that lists the images on disk is not included.  If the reader is a noob - learning the newish image/container/volume/etc api may be easier to pick up.  For example you can list images and containers via `docker image ls`/`docker container ls` and delete them via `docker image rm`/`docker container rm`

File: docs/core/docker/building-net-docker-images.md (hunk 20 above)
Comment (MichaelSimons) [reply docs/core/docker/building-net-docker-images.md]: Upon further inspection, there are the same number of steps.  The documented approach uses concepts that are more familiar to Windows users who are new to Docker.  The approach I mentioned is what someone more familiar to Docker would use.  Given these docs are more focused towards Docker noobs.  The current documentation is the preferred approach.

Please leave the doc as you have it and ignore my `docker inspect` suggestion.

File: docs/core/docker/docker-basics-dotnet-core.md (hunk 24 above)
Comment (JRAlexander) [reply docs/core/docker/docker-basics-dotnet-core.md]: Yes, that's true. I was speaking in the context of the .NET Core 2.0 changes to dotnet new. I will include a note about the changes around dotnet restore as well.

File: docs/core/docker/building-net-docker-images.md (hunk 23 above)
Comment (JRAlexander) [reply docs/core/docker/building-net-docker-images.md]: Unhderstood. Would linking to the release notes work? That way the customer will have the latest info.

File: docs/core/docker/building-net-docker-images.md (hunk 22 above)
Comment (JRAlexander) [reply docs/core/docker/building-net-docker-images.md]: Given that, it might be less confusing if we removed this altogether.  Thoughts?

File: docs/core/docker/building-net-docker-images.md (hunk 20 above)
Comment (JRAlexander) [reply docs/core/docker/building-net-docker-images.md]: Ok, thanks @MichaelSimons.  I really appreciate your thoughtful comments!

File: docs/core/docker/docker-basics-dotnet-core.md (hunk 26 above)
Comment (JRAlexander) [reply docs/core/docker/docker-basics-dotnet-core.md]: This list was initially determined and approved by several different folks during the planning of these articles. I can definitely add docker image to the list. Good catch!

File: docs/core/docker/building-net-docker-images.md (hunk 22 above)
Comment (MichaelSimons) [reply docs/core/docker/building-net-docker-images.md]: Yes I would agree that removing the size will make it less confusing.

File: docs/core/docker/building-net-docker-images.md (hunk 23 above)
Comment (MichaelSimons) [reply docs/core/docker/building-net-docker-images.md]: Yes I think that would be great.

File: docs/core/docker/building-net-docker-images.md (hunk 21 above)
Comment (JRAlexander) [reply docs/core/docker/building-net-docker-images.md]: I'm going to change the note in parentheses to reflect that thew latest tag is an alias for the SDK image. I think that might be helpful.

File: docs/core/docker/docker-basics-dotnet-core.md (hunk 24 above)
Comment (JRAlexander) [reply docs/core/docker/docker-basics-dotnet-core.md]: Actually, the linked article about dotnet restore talks about the implicit invocation quite well. Upon reflection, I think that will be a better choice. Ok with that?

File: docs/core/docker/building-net-docker-images.md (hunk 22 above)
Comment (JRAlexander) [reply docs/core/docker/building-net-docker-images.md]: Ok, as the whole point of that part was to illustrate sizes, I removed it.
